File: modules/udp-wall-clock/android/src/main/java/expo/modules/udpwallclock/UDPWallClockModule.kt
package expo.modules.udpwallclock

import android.util.Log
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.InetSocketAddress
import java.util.concurrent.ConcurrentHashMap

/**
 * Native Android module for UDP Wall Clock communication (DVB-CSS)
 * Expo Modules implementation
 */
class UDPWallClockModule : Module() {

    companion object {
        private const val TAG = "UDPWallClock"
    }

    init {
        Log.i(TAG, "🟢 UDPWallClockModule LOADED - Native Kotlin module initialized successfully!")
    }

    private val sockets = ConcurrentHashMap<Int, DatagramSocket>()
    private val listeners = ConcurrentHashMap<Int, SocketListener>()
    private var socketCounter = 0

    /**
     * Socket listener thread that receives UDP packets
     */
    private inner class SocketListener(
        private val socketId: Int,
        private val socket: DatagramSocket
    ) : Thread() {
        
        @Volatile
        private var running = true

        override fun run() {
            val data = ByteArray(32) // DVB-CSS WC packets are 32 bytes
            val packet = DatagramPacket(data, data.size)
            
            Log.d(TAG, "SocketListener started for socket $socketId")
            
            while (running && !socket.isClosed) {
                try {
                    socket.receive(packet)
                    
                    // Convert received data to hex string
                    val hexData = byteArrayToHexString(data)
                    val address = packet.address.hostAddress ?: "unknown"
                    val port = packet.port
                    
                    Log.d(TAG, "Received UDP packet: $hexData from $address:$port")
                    
                    // Send event to JavaScript
                    sendEvent("onMessage", mapOf(
                        "socketId" to socketId,
                        "data" to hexData,
                        "address" to address,
                        "port" to port
                    ))
                    
                } catch (e: Exception) {
                    if (running) {
                        Log.e(TAG, "Receive exception: ${e.message}")
                    }
                    break
                }
            }
            
            Log.d(TAG, "SocketListener stopped for socket $socketId")
        }

        fun stopListening() {
            running = false
            interrupt()
        }
    }

    override fun definition() = ModuleDefinition {
        Name("UDPWallClock")

        // Define events that can be sent to JavaScript
        Events("onMessage")

        // Create a new UDP socket
        AsyncFunction("create") { promise: Promise ->
            try {
                val socketId = ++socketCounter
                val socket = DatagramSocket(null)
                socket.reuseAddress = true
                sockets[socketId] = socket
                
                Log.d(TAG, "Socket created with id: $socketId")
                promise.resolve(socketId)
            } catch (e: Exception) {
                Log.e(TAG, "Create exception: ${e.message}")
                promise.reject("CREATE_ERROR", e.message ?: "Unknown error", e)
            }
        }

        // Bind socket to a port (0 for random port)
        AsyncFunction("bind") { socketId: Int, port: Int, promise: Promise ->
            try {
                val socket = sockets[socketId]
                if (socket == null) {
                    promise.reject("SOCKET_NOT_FOUND", "Socket $socketId not found", null)
                    return@AsyncFunction
                }
                
                socket.bind(InetSocketAddress(port))
                val boundPort = socket.localPort
                
                // Start listener thread
                val listener = SocketListener(socketId, socket)
                listeners[socketId] = listener
                listener.start()
                
                Log.d(TAG, "Socket $socketId bound to port $boundPort")
                promise.resolve(boundPort)
            } catch (e: Exception) {
                Log.e(TAG, "Bind exception: ${e.message}")
                promise.reject("BIND_ERROR", e.message ?: "Unknown error", e)
            }
        }

        // Send data to remote address
        AsyncFunction("send") { socketId: Int, hexData: String, address: String, port: Int, promise: Promise ->
            try {
                val socket = sockets[socketId]
                if (socket == null) {
                    promise.reject("SOCKET_NOT_FOUND", "Socket $socketId not found", null)
                    return@AsyncFunction
                }
                
                Log.d(TAG, "Sending to $address:$port - $hexData")
                
                val data = hexStringToByteArray(hexData)
                val packet = DatagramPacket(
                    data, 
                    data.size, 
                    InetAddress.getByName(address), 
                    port
                )
                
                // Send in background thread to avoid blocking
                Thread {
                    try {
                        socket.send(packet)
                        Log.d(TAG, "Packet sent successfully")
                    } catch (e: Exception) {
                        Log.e(TAG, "Send exception: ${e.message}")
                    }
                }.start()
                
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e(TAG, "Send exception: ${e.message}")
                promise.reject("SEND_ERROR", e.message ?: "Unknown error", e)
            }
        }

        // Close socket
        AsyncFunction("close") { socketId: Int, promise: Promise ->
            try {
                // Stop listener
                listeners[socketId]?.let { listener ->
                    listener.stopListening()
                    listeners.remove(socketId)
                }
                
                // Close socket
                sockets[socketId]?.let { socket ->
                    socket.close()
                    sockets.remove(socketId)
                }
                
                Log.d(TAG, "Socket $socketId closed")
                promise.resolve(true)
            } catch (e: Exception) {
                Log.e(TAG, "Close exception: ${e.message}")
                promise.reject("CLOSE_ERROR", e.message ?: "Unknown error", e)
            }
        }

        // Get socket info
        AsyncFunction("getSocketInfo") { socketId: Int, promise: Promise ->
            try {
                val socket = sockets[socketId]
                if (socket == null) {
                    promise.reject("SOCKET_NOT_FOUND", "Socket $socketId not found", null)
                    return@AsyncFunction
                }
                
                val info = mapOf(
                    "localPort" to socket.localPort,
                    "localAddress" to (socket.localAddress?.hostAddress ?: "0.0.0.0"),
                    "isBound" to socket.isBound,
                    "isClosed" to socket.isClosed
                )
                
                promise.resolve(info)
            } catch (e: Exception) {
                promise.reject("INFO_ERROR", e.message ?: "Unknown error", e)
            }
        }

        // Cleanup when module is destroyed
        OnDestroy {
            // Close all sockets
            listeners.values.forEach { it.stopListening() }
            listeners.clear()
            sockets.values.forEach { it.close() }
            sockets.clear()
            Log.d(TAG, "Module destroyed, all sockets closed")
        }
    }

    /**
     * Convert hex string to byte array
     */
    private fun hexStringToByteArray(s: String): ByteArray {
        val len = s.length
        val data = ByteArray(len / 2)
        var i = 0
        while (i < len) {
            data[i / 2] = ((Character.digit(s[i], 16) shl 4) + 
                          Character.digit(s[i + 1], 16)).toByte()
            i += 2
        }
        return data
    }

    /**
     * Convert byte array to hex string
     */
    private fun byteArrayToHexString(data: ByteArray): String {
        val hexArray = "0123456789ABCDEF".toCharArray()
        val hexChars = CharArray(data.size * 2)
        for (j in data.indices) {
            val v = data[j].toInt() and 0xFF
            hexChars[j * 2] = hexArray[v ushr 4]
            hexChars[j * 2 + 1] = hexArray[v and 0x0F]
        }
        return String(hexChars)
    }
}

File: modules/udp-wall-clock/ios/UDPWallClockModule.swift
import ExpoModulesCore
import Foundation
import Darwin

/**
 * Native iOS UDP module for DVB-CSS Wall Clock (CSS-WC) synchronization.
 *
 * Mirrors the Android `UDPWallClockModule` (Kotlin) API so the shared JavaScript
 * wrapper (`src/utils/NativeUDPWallClock.js`) works identically on both platforms:
 * numeric socket ids and hex-encoded payloads.
 *
 * Uses plain BSD/POSIX unicast UDP sockets (no multicast membership), so unlike
 * the `udp-multicast` module it does NOT require the
 * `com.apple.developer.networking.multicast` entitlement.
 */
public class UDPWallClockModule: Module {

  // socketId -> file descriptor
  private var sockets: [Int: Int32] = [:]
  // socketId -> running flag for the receive loop
  private var running: [Int: Bool] = [:]
  private var socketCounter = 0
  private let lock = NSLock()
  private let recvQueue = DispatchQueue(label: "cat.ccma.dialapp.udpwallclock.recv", attributes: .concurrent)

  public func definition() -> ModuleDefinition {
    Name("UDPWallClock")

    Events("onMessage")

    // Create a new UDP socket, returns the numeric socket id.
    AsyncFunction("create") { (promise: Promise) in
      let fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP)
      guard fd >= 0 else {
        promise.reject("CREATE_ERROR", "Failed to create socket (errno \(errno))")
        return
      }

      var reuse: Int32 = 1
      setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &reuse, socklen_t(MemoryLayout<Int32>.size))

      self.lock.lock()
      self.socketCounter += 1
      let socketId = self.socketCounter
      self.sockets[socketId] = fd
      self.lock.unlock()

      promise.resolve(socketId)
    }

    // Bind the socket to a local port (0 => ephemeral) and start the receive loop.
    AsyncFunction("bind") { (socketId: Int, port: Int, promise: Promise) in
      self.lock.lock()
      let fd = self.sockets[socketId]
      self.lock.unlock()

      guard let fd = fd else {
        promise.reject("SOCKET_NOT_FOUND", "Socket \(socketId) not found")
        return
      }

      var bindAddr = sockaddr_in()
      bindAddr.sin_family = sa_family_t(AF_INET)
      bindAddr.sin_port = in_port_t(UInt16(truncatingIfNeeded: port)).bigEndian
      bindAddr.sin_addr.s_addr = INADDR_ANY

      let bindResult = withUnsafePointer(to: &bindAddr) {
        $0.withMemoryRebound(to: sockaddr.self, capacity: 1) {
          Darwin.bind(fd, $0, socklen_t(MemoryLayout<sockaddr_in>.size))
        }
      }
      guard bindResult == 0 else {
        promise.reject("BIND_ERROR", "Failed to bind socket (errno \(errno))")
        return
      }

      // Resolve the actual bound port.
      var localAddr = sockaddr_in()
      var localLen = socklen_t(MemoryLayout<sockaddr_in>.size)
      let getNameResult = withUnsafeMutablePointer(to: &localAddr) {
        $0.withMemoryRebound(to: sockaddr.self, capacity: 1) { saPtr in
          getsockname(fd, saPtr, &localLen)
        }
      }
      let boundPort: Int = getNameResult == 0 ? Int(UInt16(bigEndian: localAddr.sin_port)) : port

      self.lock.lock()
      self.running[socketId] = true
      self.lock.unlock()

      self.startReceiveLoop(socketId: socketId, fd: fd)

      promise.resolve(boundPort)
    }

    // Send hex-encoded data to a remote unicast address.
    AsyncFunction("send") { (socketId: Int, hexData: String, address: String, port: Int, promise: Promise) in
      self.lock.lock()
      let fd = self.sockets[socketId]
      self.lock.unlock()

      guard let fd = fd else {
        promise.reject("SOCKET_NOT_FOUND", "Socket \(socketId) not found")
        return
      }

      guard let data = self.dataFromHexString(hexData) else {
        promise.reject("SEND_ERROR", "Invalid hex payload")
        return
      }

      var destAddr = sockaddr_in()
      destAddr.sin_family = sa_family_t(AF_INET)
      destAddr.sin_port = in_port_t(UInt16(truncatingIfNeeded: port)).bigEndian
      inet_pton(AF_INET, address, &destAddr.sin_addr)

      self.recvQueue.async {
        let sent = data.withUnsafeBytes { (rawBuffer: UnsafeRawBufferPointer) -> Int in
          withUnsafePointer(to: &destAddr) {
            $0.withMemoryRebound(to: sockaddr.self, capacity: 1) { saPtr in
              sendto(fd, rawBuffer.baseAddress, data.count, 0, saPtr, socklen_t(MemoryLayout<sockaddr_in>.size))
            }
          }
        }
        if sent < 0 {
          promise.reject("SEND_ERROR", "sendto failed (errno \(errno))")
        } else {
          promise.resolve(true)
        }
      }
    }

    // Close the socket and stop its receive loop.
    AsyncFunction("close") { (socketId: Int, promise: Promise) in
      self.closeSocket(socketId: socketId)
      promise.resolve(true)
    }

    // Return diagnostic info about the socket.
    AsyncFunction("getSocketInfo") { (socketId: Int, promise: Promise) in
      self.lock.lock()
      let fd = self.sockets[socketId]
      let isBound = self.running[socketId] ?? false
      self.lock.unlock()

      guard let fd = fd else {
        promise.reject("SOCKET_NOT_FOUND", "Socket \(socketId) not found")
        return
      }

      var localAddr = sockaddr_in()
      var localLen = socklen_t(MemoryLayout<sockaddr_in>.size)
      let getNameResult = withUnsafeMutablePointer(to: &localAddr) {
        $0.withMemoryRebound(to: sockaddr.self, capacity: 1) { saPtr in
          getsockname(fd, saPtr, &localLen)
        }
      }

      var localPort = 0
      var localAddress = "0.0.0.0"
      if getNameResult == 0 {
        localPort = Int(UInt16(bigEndian: localAddr.sin_port))
        var ipBuffer = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
        inet_ntop(AF_INET, &localAddr.sin_addr, &ipBuffer, socklen_t(INET_ADDRSTRLEN))
        localAddress = String(cString: ipBuffer)
      }

      promise.resolve([
        "localPort": localPort,
        "localAddress": localAddress,
        "isBound": isBound,
        "isClosed": false
      ])
    }

    OnDestroy {
      self.closeAll()
    }
  }

  // MARK: - Socket lifecycle

  private func closeSocket(socketId: Int) {
    lock.lock()
    let fd = sockets[socketId]
    running[socketId] = false
    sockets.removeValue(forKey: socketId)
    running.removeValue(forKey: socketId)
    lock.unlock()

    if let fd = fd {
      Darwin.close(fd)
    }
  }

  private func closeAll() {
    lock.lock()
    let ids = Array(sockets.keys)
    lock.unlock()
    for id in ids {
      closeSocket(socketId: id)
    }
  }

  // MARK: - Receive loop

  private func startReceiveLoop(socketId: Int, fd: Int32) {
    recvQueue.async { [weak self] in
      guard let self = self else { return }
      // DVB-CSS WC packets are 32 bytes; use a larger buffer to be safe.
      var buffer = [UInt8](repeating: 0, count: 2048)

      while self.isRunning(socketId) {
        var senderAddr = sockaddr_in()
        var senderLen = socklen_t(MemoryLayout<sockaddr_in>.size)

        let n = withUnsafeMutablePointer(to: &senderAddr) {
          $0.withMemoryRebound(to: sockaddr.self, capacity: 1) { saPtr in
            recvfrom(fd, &buffer, buffer.count, 0, saPtr, &senderLen)
          }
        }

        if n > 0 {
          let hex = self.hexStringFromBytes(buffer, count: n)

          var ipBuffer = [CChar](repeating: 0, count: Int(INET_ADDRSTRLEN))
          inet_ntop(AF_INET, &senderAddr.sin_addr, &ipBuffer, socklen_t(INET_ADDRSTRLEN))
          let senderIP = String(cString: ipBuffer)
          let senderPort = Int(UInt16(bigEndian: senderAddr.sin_port))

          self.sendEvent("onMessage", [
            "socketId": socketId,
            "data": hex,
            "address": senderIP,
            "port": senderPort
          ])
        } else if n <= 0 {
          // Socket closed or fatal error: exit the loop.
          break
        }
      }
    }
  }

  private func isRunning(_ socketId: Int) -> Bool {
    lock.lock()
    defer { lock.unlock() }
    return running[socketId] ?? false
  }

  // MARK: - Hex helpers (match the Android module: uppercase hex)

  private func hexStringFromBytes(_ bytes: [UInt8], count: Int) -> String {
    let hexDigits = Array("0123456789ABCDEF")
    var chars = [Character]()
    chars.reserveCapacity(count * 2)
    for i in 0..<count {
      let v = Int(bytes[i])
      chars.append(hexDigits[(v >> 4) & 0x0F])
      chars.append(hexDigits[v & 0x0F])
    }
    return String(chars)
  }

  private func dataFromHexString(_ hex: String) -> Data? {
    let chars = Array(hex)
    guard chars.count % 2 == 0 else { return nil }
    var data = Data(capacity: chars.count / 2)
    var i = 0
    while i < chars.count {
      guard let hi = chars[i].hexDigitValue, let lo = chars[i + 1].hexDigitValue else {
        return nil
      }
      data.append(UInt8((hi << 4) | lo))
      i += 2
    }
    return data
  }
}
